core/arch/aarch64/emit_utils.c

```diff
@@ -717,29 +716,33 @@ emit_indirect_branch_lookup(dcontext_t *dc, generated_code_t *code, byte *pc,
      *     x2: app's x2
      *     TLS_REG1_SLOT: app's x0 (recovered by fragment_prefix)
      */
-
     /* Spill x0. */
-    APP(&ilist, instr_create_save_to_tls(dc, DR_REG_R0, TLS_REG3_SLOT));
-    /* Load-acquire hash mask.  We need a load-acquire to ensure we see updates
-     * properly; the corresponding store-release is in update_lookuptable_tls().
+    APP(&ilist, instr_create_save_to_tls(dc, DR_REG_X0, TLS_REG3_SLOT));
+    /* Load hash mask before loading hash_table, with DMB to prevent memory-access
+     * instructions from being reordered. A corresponding store-release is in
+     * update_lookuptable_tls() to ensure a new hast_table will be written after a
+     * new hash mask write has been issued.
      */
-    /* add x1, x28 + hash_mask_offs; ldar x1, [x1]    (ldar doesn't take an offs.) */
     APP(&ilist,
-        INSTR_CREATE_add(dc, opnd_create_reg(DR_REG_X1), opnd_create_reg(dr_reg_stolen),
-                         OPND_CREATE_INT32(TLS_MASK_SLOT(ibl_code->branch_type))));
+        INSTR_CREATE_ldr(dc, opnd_create_reg(DR_REG_X1),
+                         opnd_create_base_disp(dr_reg_stolen, DR_REG_NULL, 0,
+                                               TLS_MASK_SLOT(ibl_code->branch_type),
+                                               OPSZ_8)));
+    /* and x1, x1, x2 */
+    APP(&ilist,
+        INSTR_CREATE_and(dc, opnd_create_reg(DR_REG_X1), opnd_create_reg(DR_REG_X1),
+                         opnd_create_reg(DR_REG_X2)));
+    /* dmb ishld to prevent memory-access instructions from being reordered across the
+     * barrier. DMB results generally in better performances compared to LDAR in the case.
+     */
     APP(&ilist,
-        INSTR_CREATE_ldar(dc, opnd_create_reg(DR_REG_X1),
-                          OPND_CREATE_MEMPTR(DR_REG_X1, 0)));
+        instr_create_0dst_1src(dc, OP_dmb, OPND_CREATE_INT(/* DR_DMB_ISHLD */ 9)));
```

Comment:
We'd prefer to use INSTR_CREATE_macros: ARM and x86 are complete but it seems AArch64 is missing some.  Could you add the ones you need?  IIRC they are manually done instead of automated like the ARM ones are.

Reply:
Got it. I have created INSTR_CREATE_dmb macro for AArch64 in the 2nd commit.